embulk-util-throttle-helper/src/main/java/org/embulk/util/throttle_helper/QuotaRateLimiter.java

```diff
@@ -0,0 +1,157 @@
+package org.embulk.util.throttle_helper;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.Uninterruptibles;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * A rate limiter that allows a specific number of operations happen within
+ * a time window and its rate limit is only reset at the beginning of the
+ * next time window.
+ * <p>
+ * The time window is currently in second resolution to make sure unused permit
+ * doesn't get lost in situation when a thread, that must sleep in more than 2
+ * time windows, wakes up before all other threads, and reset the limits.
+ *
+ * @author instcode
+ * @author thangnc
+ */
+@ThreadSafe
+public class QuotaRateLimiter
+{
+    /**
+     * This is to prevent multiple threads from updating {@link #remainingLimit} &
+     * {@link #lastRefreshTime} wrongly
+     */
+    private Lock lock = new ReentrantLock();
+
+    /**
+     * The number of permits per the given time window
+     */
+    private int rateLimit;
```

Comment:
I suggest to make it `final` explicitly.

Reply:
Sure. Will update.